### PR TITLE
Fixes #38477 - Allow rolling CVs in arbitrary environments

### PR DESCRIPTION
**Pull request by quba42**
Initial state for rolling content views with arbitrary lifecycle environments.

Things that are missing from this "Draft" PR:

- Tests for the API controller and backend actions.
- Checking the new API parameter for consistent behavior with similar existing parameters.

In general, there are various possible combinations of API parameters and situations I have not yet tested.

### CODE REVIEW

**Comment by quba42** [comment]
Note, there is also already a accompanying documentation PR: https://github.com/theforeman/foreman-documentation/pull/3902

**Comment by quba42** [comment]
I am not currently aware of any known issues with the latest state. I do still plan to add more test coverage.

**Comment by quba42** [comment]
The latest state still has a known issue, where automatic smart proxy syncs when content on the server is updated are scheduled twice in a row instead of just once.

**Comment by quba42** [comment]
I have now significantly expanded the actions test coverage, especially for the `ContentView::Update` action.
With this I consider this ready for final review and am undrafting the PR.

Note that https://github.com/Katello/katello/pull/11433 should be merged first, and then this PR should be rebased on top.

File: app/controllers/katello/api/v2/content_views_controller.rb [hunk 1]
```diff
@@ -299,12 +301,17 @@ def view_params
...
+      result = {}
+      result = params.require(:content_view).permit(*attrs).to_h unless action_name == "update" && @content_view.rolling? && params[:content_view].empty?
       # sanitize repository_ids to be a list of integers
       result[:repository_ids] = result[:repository_ids].compact.map(&:to_i) if result[:repository_ids].present?
       result
     end
 
+    def sanitized_environment_ids
+      params[:environment_ids]&.compact&.map(&:to_i)
+    end
```
[the remark this answers is not in the document]
Comment (quba42) [reply app/controllers/katello/api/v2/content_views_controller.rb:315]: This line is inspired by https://github.com/Katello/katello/blob/master/app/controllers/katello/api/v2/content_views_controller.rb#L304, not sure if this changes anything.

**Comment by quba42** [comment]
Reminder that there is also a docs PR to go with this: https://github.com/theforeman/foreman-documentation/pull/3902

File: app/controllers/katello/api/v2/content_views_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (m-bucher) [reply app/controllers/katello/api/v2/content_views_controller.rb:315]: Looking at the code where we use the result of `sanitized_environment_ids()`, most of them acknowledge the value might be `nil`.
The only code where this could theoretically be problematic is in `Actions::Katello::ContentView::AddRollingRepoClone`, where it is directly used to query the DB.
There having it being `nil` will currently result in the following query being made:
```
::Katello::KTEnvironment.where(id: nil)
```
being translated to `SELECT * FROM katello_environments WHERE id IS NULL`. Instead of the query we actually want, which is `SELECT * FROM katello_environments WHERE is IN ()`.
Practically, both queries return no entries, with `id` being the Primary Key of the DB-table and therefore not being allowed to be `NULL` for any DB-entry.

If you make sure `sanitized_environment_ids()` always returns an `Array`-type object, then you could avoid checking the value for being `nil` in any occurrences.

## tldr
Your call @quba42, should be fine either way :grin:

**Comment by quba42** [comment]
> While I still plan on getting team reviews on this soon, are there any issues with this merging after Katello 4.18 branches (so for 4.19)? We're slowing down feature development in Katello after next week for extended stabilization and bug fixing. The target branch date is August 12th: https://community.theforeman.org/t/foreman-3-16-schedule-and-planning/43345

I have no skin in this game (the feature extension landing in 4.18 vs. 4.19). I do have a preference for the following related PRs to make it into 4.18:

- https://github.com/Katello/katello/pull/11438
- https://github.com/Katello/katello/pull/11433

It looks like you are already aware of those and one has already received some review and the ball is back in my court.

The reasons those would be helpful are:

a) They are more bugfix than feature and bugs are not nice.
b) There are minor merge conflicts between this PR and the other two. Once they are merged, I can re-base this PR, resolve the conflicts, and forget about them. :wink:

**Comment by quba42** [comment]
I will be on PTO for the next two weeks. During this time @m-bucher will be in charge of this PR.

**Comment by quba42** [comment]
Summary of current state:

- I am not aware of any issues with functionality.
- I am happy with the test coverage we added to go with the new feature.
- I have yet to deal with the two sourcery-ai suggestions above.

=> IMHO this work is very close to complete.

We will not merge before Katello 4.18 is branched, but we are ready for final reviews.

**Review by ianballou — COMMENTED** [review]
Initial testing is solid, I haven't run into any issues yet. I've created a rolling CV with a few repositories and a couple life cycle environments. I've synced those environments to a smart proxy, and the content appeared. I also tested uploading content to a repo in my rolling CV, syncing it up to the smart proxy, and verifying that the content is where it needs to be.

We should find more corner cases to test and proceed to code review.

File: app/lib/actions/katello/content_view/update.rb [hunk 2]
```diff
@@ -28,12 +30,33 @@ def plan(content_view, content_view_params)
...
+
+              ::Katello::KTEnvironment.where(id: environment_ids_to_remove).each do |environment|
+                plan_action(RemoveFromEnvironment, content_view, environment)
+              end
+            end
 
-            plan_action(AddRollingRepoClone, content_view, repo_ids_to_add) if repo_ids_to_add.any?
-            plan_action(RemoveRollingRepoClone, content_view, repo_ids_to_remove) if repo_ids_to_remove.any?
+            plan_action(AddRollingRepoClone, content_view, repo_ids_to_add, environment_ids_for_new_repos) if repo_ids_to_add.any?
+            plan_action(RemoveRollingRepoClone, content_view, repo_ids_to_remove, retained_environment_ids) if repo_ids_to_remove.any?
```
Comment (ianballou) [inline app/lib/actions/katello/content_view/update.rb]: It would be nice to split this big special handling of rolling content views into smaller methods so the rubocop rules don't need to be disabled.

Maybe something along the lines of:

```
 if content_view.rolling?
    # Calculate repository changes
    repo_changes = calculate_repo_changes(content_view,
  content_view_params)

    # Calculate environment changes  
    env_changes = calculate_env_changes(...)

    # Apply environment changes
    apply_environment_changes(...) if env_changes

    # Apply repository changes
    apply_repository_changes(...)
  end
```

Comment (quba42) [reply app/lib/actions/katello/content_view/update.rb]: I refactored to implement a variant of this proposal. I kept it as a separate commit for now, but I am happy to squish when needed. The rubocop disables are no longer needed.

The test coverage I added remained green with the refactor, so either it did not break anything or my tests are bad.
Manual re-testing is going well so far, so I am hoping for the former. :relaxed:

File: webpack/scenes/ContentViews/Details/ContentViewInfo.js [hunk 3]
```diff
@@ -35,16 +46,49 @@ const ContentViewInfo = ({ cvId, details }) => {
...
+    if (environments?.length) {
+      setSelectedEnvs(environments);
+    }
+  }, [environments, setSelectedEnvs]);
+  useMemo(() => {
+    if (rolling && !environmentPathLoading) {
+      const { results } = environmentPathResponse || {};
+      return results.map(result => result.environments).flatten();
+    }
+    return [];
```
[the remark this answers is not in the document]
Comment (quba42) [reply webpack/scenes/ContentViews/Details/ContentViewInfo.js]: We dropped useMemo and re-tested and found no ill effects.

Our best guess is that we thought we needed this from a mixture of not clearing our browser cache during development (making environment loading appear broken when it wasn't) and seeing it used elsewhere.

File: webpack/scenes/ContentViews/Create/CreateContentViewForm.js [hunk 4]
```diff
@@ -100,6 +121,13 @@ const CreateContentViewForm = ({ setModalOpen }) => {
     }
   }
 
+  useMemo(() => {
+    if (rolling && !environmentPathLoading) {
+      const { results } = environmentPathResponse || {};
+      return results.map(result => result.environments).flatten();
+    }
+    return [];
```
[the remark this answers is not in the document]
Comment (quba42) [reply webpack/scenes/ContentViews/Create/CreateContentViewForm.js]: Same as the above use of useMemo.

File: app/lib/actions/katello/content_view/update.rb (hunk 2 above)
Comment (quba42) [reply app/lib/actions/katello/content_view/update.rb]: I squished things together, but I still have the previous version on local branch if needed.

File: app/controllers/katello/api/v2/content_views_controller.rb [hunk 5]
```diff
@@ -21,6 +21,7 @@ class Api::V2::ContentViewsController < Api::V2::ApiController
       param :description, String, :desc => N_("Description for the content view")
       param :repository_ids, Array, :desc => N_("List of repository ids")
       param :component_ids, Array, :desc => N_("List of component content view version ids for composite views")
+      param :environment_ids, Array, :desc => N_("List of lifecycle environments for rolling content views")
```
[the remark this answers is not in the document]
Comment (quba42) [reply app/controllers/katello/api/v2/content_views_controller.rb:24]: Right now, if you supply `environment_ids` to a non rolling content view, then those environment IDs will simply be ignored. I am fine with leaving this behavior. I am also fine with having validation to tell the user they are doing something meaningless in those cases.

I would not add validation to check the integrity of any list provided, since we are not doing this for any similar parameters, so adding it just for this parameter would make the API less consistent, and starting to add such validation for all relevant parameters is clearly out of scope of this PR.

File: app/lib/actions/katello/content_view/add_rolling_repo_clone.rb [hunk 6]
```diff
@@ -2,36 +2,28 @@ module Actions
...
+          ::Katello::KTEnvironment.where(id: environment_ids).each do |environment|
+            concurrence do
+              ::Katello::Repository.where(id: repository_ids).each do |repository|
+                sequence do
+                  clone = content_view.get_repo_clone(environment, repository).first
+                  if clone.nil?
+                    clone = repository.build_clone(content_view: content_view, environment: environment)
+                    clone.save!
+                  end
+                  plan_action(RefreshRollingRepo, clone, false)
```
[the remark this answers is not in the document]
Comment (quba42) [reply app/lib/actions/katello/content_view/add_rolling_repo_clone.rb]: At first glance this looks like a sensible improvement to me.

[the remark this answers is not in the document]
Comment (quba42) [reply app/lib/actions/katello/content_view/add_rolling_repo_clone.rb]: Applied with the latest changes.

**Comment by quba42** [comment]
Summary of latest state:

- I have applied some suggestions from the latest sourcery-ai review that I found sensible.
- After discussion with @m-bucher we decided we would like to ignore the remaining suggestions from the sourcery-ai reviews, either because we disagree with them, or because we feel they are demanding a level of additional work that is not warranted by the expected gain in quality. If any human reviewers disagree, feel free to speak up! :smile: 
- I also fixed a race condition in the `RemoveRollingRepoClone` action (some things were being done concurrently which need to be strictly sequential). I chanced upon the failure case (which only occurs sometimes) by chance. Did cost me some time to figure out.

=> My preference is now to try to finalize and merge this work without too many more rounds of review. Which obviously should not preclude any good suggestions that are still found during further review. :wink:

**Review by ianballou — COMMENTED** [review]
Initial retesting looking good, just need to test with actual clients and then do a final code review.

File: app/lib/actions/katello/content_view/create.rb [hunk 7]
```diff
@@ -2,20 +2,25 @@ module Actions
   module Katello
     module ContentView
       class Create < Actions::Base
-        def plan(content_view)
+        def plan(content_view, environment_ids)
```
Comment (ianballou) [inline app/lib/actions/katello/content_view/create.rb]:
```suggestion
        def plan(content_view, content_view_params, environment_ids = [])
```
This way, we could remove the need for save navigation down below. The suggestion here also applies to the other content view actions.

These environment IDs currently look like they're a general option, but they really only apply to rolling CVs. Perhaps they could be renamed to `rolling_environment_ids`? I don't envision environment IDs being used for other CV types.

File: app/controllers/katello/api/v2/content_views_controller.rb (hunk 5 above)
Comment (ianballou) [inline app/controllers/katello/api/v2/content_views_controller.rb:24]: Related to the sourcery review - silently ignoring incompatible options is something I think we need to move away from. We do it when someone tries to update the type, which isn't great since the user may think it actually did something.

Similar to my other comment, I'd like the environment IDs to be separated out via validations and naming. CVs are a bit confusing - I could see someone in the future glancing at this and missing the direct connection between CV env IDs and rolling CVs.

**Review by ianballou — CHANGES_REQUESTED** [review]
Some code comments mostly about how the environment_ids are used:

File: app/controllers/katello/api/v2/content_views_controller.rb (hunk 5 above)
Comment (quba42) [reply app/controllers/katello/api/v2/content_views_controller.rb:24]: Adding validation that tells users not to add meaningless information to their API requests, rather than quietly ignoring the meaningless part of the request makes a lot of sense to me. This is the part of the sourcery-ai review, that I did not implement immediately, that I was most sympathetic to.

I will add validation that throws an error if users supply environment_ids and the CV is not a rolling one.

File: app/lib/actions/katello/content_view/create.rb (hunk 7 above)
Comment (quba42) [reply app/lib/actions/katello/content_view/create.rb]: Defaulting `environment_ids` to the empty list makes sense on the create action. For the update action nil/not provided and empty list have different meanings ("leave unchanged" vs "remove all existing environments").

I am against renaming the parameter `rolling_environment_ids`. I think it is instructive to look at the following screen shot from the API docs here:
image

- `component_ids` which only has meaning for composite content views is not named `composite_component_ids`
- `repository_ids` has meaning for both normal and rolling, but not for composite, how would we even name that?
- Some users might think their is a type of special "rolling environments" that is somehow different or separate from normal lifecycle environments.
- In the purely hypothetical scenario where we added a feature in the future such that normal content views could consume a list of `environment_ids`, would we then want both `rolling_environment_ids` and `environment_ids`?

=> We have some parameters that are overloaded for the different types and some that are only meaningful for one or two of the three types. I think the reasonable compromise is to name them as "general options" make sure the description says "for rolling content views" or "for composite views", and then have validation that immediately tells users (who don't read the API docs) when they are supplying a parameter to their API call that does not make sense for the type of content view they are operating on.

File: app/controllers/katello/api/v2/content_views_controller.rb (hunk 5 above)
Comment (quba42) [reply app/controllers/katello/api/v2/content_views_controller.rb:24]: I added the validation with the latest push. I tried to make it similar to other existing validation in the controller, but I am not sure if there are better choices for the error type or general style of the added code?

File: app/lib/actions/katello/content_view/create.rb (hunk 7 above)
Comment (quba42) [reply app/lib/actions/katello/content_view/create.rb]: Applied the `environment_ids = []` fallback suggestion.

File: app/lib/actions/katello/content_view/update.rb [hunk 8]
```diff
@@ -2,7 +2,7 @@ module Actions
   module Katello
     module ContentView
       class Update < Actions::EntryAction
-        def plan(content_view, content_view_params)
+        def plan(content_view, content_view_params, environment_ids)
```
Comment (ianballou) [inline app/lib/actions/katello/content_view/update.rb:5]:
```suggestion
        def plan(content_view, content_view_params, environment_ids = [])
```
For consistency with CV create

File: test/actions/katello/content_view_test.rb [hunk 9]
```diff
@@ -537,7 +551,7 @@ class CreateTest < TestBase
 
     it 'plans' do
       content_view.expects(:save!)
-      plan_action(action, content_view)
+      plan_action(action, content_view, nil)
```
Comment (ianballou) [inline test/actions/katello/content_view_test.rb]: The optional argument lets us keep these `nil` params out now. Same comment for the other tests.

Comment (ianballou) [reply test/actions/katello/content_view_test.rb]: Feel free to leave them, but I suggested the optional argument partially so fewer tests had to be modified.

File: app/controllers/katello/api/v2/content_views_controller.rb (hunk 5 above)
Comment (ianballou) [reply app/controllers/katello/api/v2/content_views_controller.rb:24]: I need to test this  validation, but the implementation makes sense to me :+1:

File: app/lib/actions/katello/content_view/update.rb (hunk 8 above)
Comment (quba42) [reply app/lib/actions/katello/content_view/update.rb:5]: I don't think we can use this on the update because `nil` and `[]` have different meanings here. Not providing the `environment_ids` parameter (`nil`) will cause the environments associated with the rolling CV to remain unchanged, while providing an empty list will cause all environments to be disassociated from the rolling CV (providing the empty list = "I don't want any environments to be associated with this rolling CV")

I will run some tests tomorrow to be sure.

Comment (ianballou) [reply app/lib/actions/katello/content_view/update.rb:5]: In that case, perhaps both should be `nil`? I'm okay with whatever default value makes the most sense.

Comment (quba42) [reply app/lib/actions/katello/content_view/update.rb:5]: I actually think it makes sense to have the [] default in the create action, but not in the update action. The reason is that this makes the actions more internally consistent. In the create action both `repository_ids` and `environment_ids` are now assumed to not be `nil`. While within update both of them can be `nil` and are handled accordingly. In the create action `nil` has no functional meaning for either list. In the update action `nil` vs `[]` has a destinct well defined meaning for both lists.

File: test/actions/katello/content_view_test.rb (hunk 9 above)
Comment (quba42) [reply test/actions/katello/content_view_test.rb]: I have dropped several nil and [] for the create actions in the test, but not for the update action, which still needs the distinction.

**Comment by quba42** [comment]
I have responded to all suggestions/threads. I believe we are back in a potentially mergable state.

I am glad we added in the validation. I feel like I should have added it when the AI reviewer suggested it, but chose to be lazy instead. :wink:

**Review by ianballou — APPROVED** [review]
ACK pending the tests passing, thanks!
